src/components/item/Item.js

```diff
@@ -1,85 +1,139 @@
-import React, { useState } from 'react';
+import React, { useEffect, useState } from 'react';
 import { useStore } from 'store/index';
 
-const Item = props => {
+const Item = ({ product }) => {
+    const { id, img, name, price, size: inventory } = product;
     const { dispatch, state } = useStore();
-    const { size } = state;
+    const [size, setSize] = useState();
+    const [quantity, setQuantity] = useState(1);
+    const [formErrors, setFormErrors] = useState({
+      currentlyInCart: false,
+      currentlyInWishlist: false,
+      size: false,
+      quantity: false,
+      outOfStock: false,
+      insufficientStock: false
+    });
 
-    const [numberOfItems, setNumberOfItems] = useState({ number: 1, id: null });
+    const handleAddToCartClick = () => {
+      const errors = {
+        quantity: quantity < 1,
+        size: !size,
+        currentlyInCart: !!state.cart.find(({ id: productId }) => id === productId),
+        outOfStock: inventory[size] < 1,
+        insufficientStock: quantity > inventory[size]
+      };
 
-    const pickSizeToColor = sizeToColor => {
-        if (+size.id === props.product.product_id && sizeToColor === size.size) {
-            return 'size';
-        }
+      const hasError = Object.values(errors).reduce((result, val) => val ? true : result, false);
+
+      setFormErrors(errors);
+
+      if (!hasError) {
+        dispatch({
+          type: 'addItemToCart',
+          payload: { size, quantity, product }
+        });
+      }
     };
 
-    const handleWishClick = e => {
-        let updatedWishList = state.wishList;
-        for (let i = 0; i < updatedWishList.length; i++) {
```

Comment:
`let`, modifying variables outside scope in `for` loops are telltale signs of imperative code. it's not "worse" or "better" but paradigmatically does not really fit with how react works.